lib/http_aws_sigv4.c

```diff
@@ -66,6 +68,186 @@ static void sha256_to_hex(char *dst, unsigned char *sha, size_t dst_l)
   }
 }
 
+/*
+ * Decide in an encoding-independent manner whether a character in an
+ * URL must be escaped. The same criterion must be used in strlen_uri()
+ * and strcpy_uri().
+ */
+static bool urlchar_needs_escaping(int c, bool in_query)
+{
+  return !((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') ||
+           (c >= '0' && c <= '9') || c == '-' || c == '.' || c == '_'
+           || c == '~' ||
+           (!in_query && c == '/') ||
+           (in_query && c == '&'));
+}
+
+
+/*
+ * encode_query_len() returns the length of the given URI
+ * this fucntion is a modified version of strlen_url in urlapi.c
+ */
+static size_t encode_query_len(const char *url, bool in_query)
+{
+  const unsigned char *ptr = (unsigned char *)url;
+  size_t newlen = 0;
+  int count_equal = 0;
+
+  if(!*ptr)
+    return 0;
+
+  for(; *ptr; ptr++) {
+    switch (*ptr) {
+    case '=':
+      if(!in_query || count_equal++)
+        newlen += 2;
+      newlen++;
+      break;
+    default:
+      if(urlchar_needs_escaping(*ptr, in_query))
+        newlen += 2;
+      newlen++;
+    }
+  }
+  if(in_query && !count_equal)
+    newlen += 1;
+  return newlen;
+}
+
+static char *decode_query_cpy(char *optr, char *iptr, bool in_query)
+{
+  for(;
+       *iptr;
+       iptr++) {
+    if(in_query && *iptr == '&') {
+      ++iptr;
+      goto out;
+    }
+    if(*iptr == '%' && iptr[1] && iptr[2]) {
+      char num_str[3] = {iptr[1], iptr[2], 0};
+      long ascii_code;
+
+      ascii_code = strtoul(num_str, NULL, 16);
+      if(ascii_code) {
```

Comment:
`%00` is silently just ignored here. Is that on purpose? Isn't that an error if found in the URL?

What about other control codes like `%0a` or `%0d` ? They will certainly mess up the output.